Lib/asyncio/streams.py

```diff
@@ -331,9 +339,31 @@ def transport(self):
 
     def write(self, data):
         self._transport.write(data)
+        return self._fast_drain()
 
     def writelines(self, data):
         self._transport.writelines(data)
+        return self._fast_drain()
+
+    def _fast_drain(self):
+        # The helper tries to use fast-path to return already existing complete future
+        # object if underlying transport is not paused and actual waiting for writing
+        # resume is not needed
+        if self._reader is not None:
+            # this branch will be simplified after merging reader with writer
+            exc = self._reader.exception()
+            if exc is not None:
+                fut = self._loop.create_future()
+                fut.set_exception(exc)
+                return fut
+        if not self._transport.is_closing():
+            if self._protocol._connection_lost:
+                fut = self._loop.create_future()
+                fut.set_exception(ConnectionResetError('Connection lost'))
+                return fut
+            if not self._protocol._paused:
+                return self._complete_fut
+        return self._loop.create_task(self.drain())
```

Comment:
The idea is to avoid creating task in majority of the cases, e.g.:

In the stream `__init__`:

```python
   self._nop = self.loop.create_future()
   self._nop.set_result(None)
```

and then in `_fast_drain`:

```python
   if not self._paused:
        return self._nop
```

Reply:
But I do almost the same a few lines above.
There is no `writer._paused` attribute but `writer._protocol._paused` one (which has undefined state if the protocol is closing).
That's why I have slightly more comlex logic that you mentioned in a comment but the functionality is the same.

Reply:
Ah, I see, never mind.  Could you please add a comment before `return self._complete_fut` that it's a "fast path"?

Reply:
Done